File: Easy/problem1/main.go
package main

// Solution for Memory only

/*
Runtime: 28 ms, faster than 32.44% of Go online submissions for Two Sum.
Memory Usage: 2.9 MB, less than 99.94% of Go online submissions for Two Sum.
16/08/2020
DD/MM/YY
*/

func in(target int, list []int) bool {
	for _, i := range list {
		if target == i {
			return true
		}
	}

	return false
}

func count(target int, list []int) int {
	total := 0
	for _, i := range list {
		if i == target {
			total++
		}
	}

	return total
}

func indice(target int, list []int) []int {
	indices := make([]int, 0, 0)
	for i := 0; i < len(list); i++ {
		if list[i] == target {
			indices = append(indices, i)
		}
	}

	return indices
}

func indexOf(target int, list []int) int {
	for k, v := range list {
		if target == v {
			return k
		}
	}

	return -1
}

func twoSum(nums []int, target int) []int {
	for _, i := range nums {
		if in(target-i, nums) {
			if count(target-i, nums) >= 2 {
				indices := indice(target-i, nums)

				return []int{indices[0], indices[1]}
			} else if indexOf(i, nums) == indexOf(target-i, nums) {
				continue
			} else {
				return []int{indexOf(i, nums), indexOf(target-i, nums)}
			}
		}
	}

	return nil
}

File: easy/problem1/main.cpp
/**
 * Runtime: 164 ms, faster than 18.73% of C++ online submissions for Two Sum.
 * Memory Usage: 9.3 MB, less than 81.47% of C++ online submissions for Two Sum.
 * 02/21/2020
 * MM/DD/YYYY
 */

#include <vector>

bool contains(int target, std::vector<int>& list) {
    for (const int &i : list) {
        if (target == i) {
            return true;
        }
    }

    return false;
}

int count(int target, std::vector<int>& list) {
    int total = 0;

    for (const int &i : list) {
        if (i == target) {
			total++;
		}
	}
    
	return total;
}

int indexOf(int target, std::vector<int>& list) {
    for (unsigned int i = 0; i < list.size(); i++) {
        if (target == list[i]) {
            return i;
        }
    }

    return -1;
}

std::vector<int> indice(int target, std::vector<int>& list) {
    std::vector<int> indices;
    
    for (unsigned int i = 0; i < list.size(); i++) {
        if (list[i] == target) {
            indices.push_back(i);
        }
    }

    return indices;
}

class Solution {
    public:
    std::vector<int> twoSum(std::vector<int>& nums, int target) {
        for (const int &i : nums) {
            if (contains(target-i, nums)) {
                if (count(target-i, nums) >= 2) {
                    std::vector<int> indices = indice(target-i, nums);
                    std::vector<int> res = {indices[0], indices[1]};

                    return res;
                } else if (indexOf(i, nums) == indexOf(target-i, nums)) {
                    continue;
                } else {
                    std::vector<int> res = {indexOf(i, nums), indexOf(target-i, nums)};

                    return res;
                }
            }
        }
        std::vector<int> res = {};

        return res;
    }
};
